tools/tpm2_encryptdecrypt.c

```diff
@@ -81,28 +78,55 @@ static bool encrypt_decrypt(ESYS_CONTEXT *ectx, TPM2B_IV *iv_in) {
         return false;
     }
 
-    TSS2_RC rval = Esys_EncryptDecrypt2(ectx, ctx.object.context.tr_handle,
+    UINT16 data_offset = 0;
+    bool result = true;
+    bool append = false;
+    while (ctx.input_data_size > 0) {
+
+        ctx.data.size =
+            ctx.input_data_size > TPM2_MAX_DIGEST_BUFFER ?
+                    TPM2_MAX_DIGEST_BUFFER : ctx.input_data_size;
+
+        memcpy(ctx.data.buffer, &ctx.input_data[data_offset], ctx.data.size);
+
+        TSS2_RC rval = Esys_EncryptDecrypt2(ectx, ctx.object.context.tr_handle,
+                          shandle1, ESYS_TR_NONE, ESYS_TR_NONE,
+                          &ctx.data, ctx.is_decrypt, ctx.mode, iv_in,
+                          &out_data, &iv_out);
+        if (tpm2_error_get(rval) == TPM2_RC_COMMAND_CODE) {
+            version = 1;
+            rval = Esys_EncryptDecrypt(ectx, ctx.object.context.tr_handle,
                       shandle1, ESYS_TR_NONE, ESYS_TR_NONE,
-                      &ctx.data, ctx.is_decrypt, ctx.mode, iv_in,
+                      ctx.is_decrypt, ctx.mode, iv_in, &ctx.data,
                       &out_data, &iv_out);
-    if (tpm2_error_get(rval) == TPM2_RC_COMMAND_CODE) {
-        version = 1;
-        rval = Esys_EncryptDecrypt(ectx, ctx.object.context.tr_handle,
-                  shandle1, ESYS_TR_NONE, ESYS_TR_NONE,
-                  ctx.is_decrypt, ctx.mode, iv_in, &ctx.data,
-                  &out_data, &iv_out);
-    }
-    if (rval != TPM2_RC_SUCCESS) {
-        if (version == 2) {
-            LOG_PERR(Esys_EncryptDecrypt2, rval);
+        }
+        if (rval != TPM2_RC_SUCCESS) {
+            if (version == 2) {
+                LOG_PERR(Esys_EncryptDecrypt2, rval);
+            } else {
+                LOG_PERR(Esys_EncryptDecrypt, rval);
+            }
+            return false;
+        }
+
+        if (!append) {
+            result = files_save_bytes_to_file(ctx.out_file_path,
+                out_data->buffer, out_data->size);
         } else {
-            LOG_PERR(Esys_EncryptDecrypt, rval);
+            result = files_append_bytes_to_file(ctx.out_file_path,
+                out_data->buffer, out_data->size);
```

Comment:
weird way to do this. Open the file in rw mode and keep the FILE * open the whole time and just write it out.

Reply:
using files_write_bytes routine instead.